File: src/single-flight-lock.js
'use strict';

// Node-native equivalent of agent-manager-common.sh's acquire_single_flight_lock()/
// release_single_flight_lock() -- same real flock(2) mutex, same lockfile
// (instances/.pipeline-single-flight.lock), fully interoperable with the bash version
// worker-1/reviewer still use (confirmed live 2026-08-22: a Node-held lock correctly
// blocks a separate bash `exec 200>file; flock 200` process, and releases it the instant
// Node closes its own fd -- flock(2) locks are owned by the OPEN FILE DESCRIPTION, not
// the process, so this works precisely because both sides open the SAME file).
//
// Built to fix a real, observed problem (2026-08-22, Grimmethy: "build [a real plan/
// implement lock split] now"): local-draft.js's draftTask() runs an adhoc/research
// task's PLAN pass through the local model (genuinely needs this lock) then unconditionally
// bypasses to a real Claude call for IMPLEMENT (never touches the local GPU at all) --
// but the OLD bash-level lock in local-worker.sh had to choose ONE lock decision for the
// whole `node local-draft.js` call, so it either protected the plan pass and then kept
// holding the lock through the long Claude call too (needless cross-lane contention,
// confirmed live: real "Ollama request timed out after ~130s" plan-stage failures caused
// by exactly this), or skipped the lock entirely and left the plan pass unprotected
// (the earlier, WORSE bug this session already found and reverted). This module lets
// draftTask() hold the lock ONLY around the specific local-model call that actually
// needs it, released immediately afterward -- local-worker.sh no longer does any
// lock-wrapping of its own for the draft stage at all (see its own updated comment).
//
// A held lock here MUST be released -- an uncaught throw between acquire() and release()
// would leak the fd and deadlock every other lane forever. Callers should always use
// withLock() below rather than acquire()/release() directly unless they have a very
// specific reason not to (release() is still exported for that rare case).
//
// ---------------------------------------------------------------------------------------
// 2026-08-31: two fairness/correctness fixes, after task
// adhoc-brain-dump-agent-manager-blocked-task-done-task-bra-1788142124203 made 19 draft
// attempts over ~11.5h and blocked every time on a 600s lock timeout. Root cause was
// TWO bugs (see brain dump "GPU single-flight lock: replace the unfair flock(2) mutex
// with a FIFO ticket lock"):
//
//   (1) SELF-DEADLOCK / non-reentrancy. draftAdhocBranch() wraps a tier in
//       maybeLocked(...) -> withLock(dir, fn, MODEL) which holds an fd on the lockfile
//       for the WHOLE tier. Inside fn, local-tool-client.js's runPlanWithTools() takes
//       withLock(dir, ..., MODEL) again PER TURN -- same process, same key, but a
//       DIFFERENT open file description -> flock(2) blocks the inner call against the
//       outer one for the full timeout, every turn. Empirically confirmed. Fixed here
//       with per-process reentrancy: a lock already held by THIS process for a given key
//       is re-granted immediately (depth-counted), released only when the outermost
//       holder releases.
//
//   (2) UNFAIRNESS. flock(2) gives no ordering guarantee -- waiters wake arbitrarily --
//       so under sustained contention a third consumer starves. In the incident,
//       worker-reasoning's tier lost the acquire race to worker-1 + the reviewer for
//       600s straight, every attempt. Fixed here with a FIFO ticket queue: a waiter
//       drops a lexically-time-ordered marker in instances/.pipeline-single-flight-queue
//       [.<model>]/ and only attempts the real flock once it holds the OLDEST live
//       ticket (dead-pid / stale tickets are swept). The final flock(2) is still the
//       real mutual-exclusion primitive; the queue just decides who attempts it next.
//
// NOT yet migrated: scripts/agent-manager-common.sh (review-runner) and
// python/dashboard/single_flight_lock.py (dashboard chat/Discuss) do NOT participate in
// the ticket queue -- they still contend on the final flock directly, so they can
// occasionally cut ahead of the JS worker FIFO. Acceptable for now: both hold the lock
// briefly (one review vote / one chat turn) and release. Full cross-language ticket
// participation is a fast-follow (same brain dump). All three twins still share the same
// final lockfile, so mutual exclusion and the Discuss-priority backoff are unaffected.
// Python's acquire() DID get bounded-wait parity here (it was still unbounded).

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { execFileSync } = require('child_process');

const LOCK_CHILD_FD = 3; // arbitrary -- just needs to not collide with the child's own 0/1/2

// Per-model locking (2026-08-25, Grimmethy: "boost throughput" -- see nullboiler's
// ConcurrencyConfig.per_role for the reference shape this borrows). Originally ONE
// lockfile for every local-model call regardless of which model it targeted, which meant
// a long call on the reasoning model (e.g. a big arch_review) blocked a completely
// unrelated cheap-model call (e.g. brain_dump_sort's qwen2.5:3b) for its entire duration
// even after OLLAMA_MAX_LOADED_MODELS=2 let both models stay resident in VRAM at once --
// confirmed live: this was the literal "worker-1 and reasoning are taking turns" report.
// `key` (typically the resolved model name, e.g. labelFor(task) for a local call) scopes
// the lockfile to that specific model, so two calls against DIFFERENT models no longer
// serialize against each other, while two calls against the SAME model still correctly
// do (same key -> same lockfile -> same flock). Omitting `key` preserves the exact
// original global lockfile name -- every caller not yet migrated to pass a key keeps
// today's behavior unchanged.
//
// UPDATE, 2026-08-25, same day: this comment used to say bash's own
// acquire_single_flight_lock/release_single_flight_lock (agent-manager-common.sh, used by
// review-runner.sh) was "intentionally NOT migrated" and that the gap for the common
// (no cheap-model-profile) case was "a known, accepted, low-frequency risk" because
// "Ollama's own llama-server still serializes same-model requests at its own -np 1 slot
// regardless." Confirmed live, same day, that assumption was wrong: reviewer (still on
// the old global lockfile) and worker-1 (already on this module's per-model one) hitting
// the SAME default model produced a real, repeating "timed out waiting for llama-server
// to start: context canceled" cycle in Ollama's own log -- concurrent LOAD attempts
// collide and get cancelled, not just generation slots, so this was the common case
// racing on every single review, not a rare edge case. agent-manager-common.sh's
// acquire_single_flight_lock now also accepts an optional key (same lockFileName scheme
// as here, kept in sync by hand) and review-runner.sh passes it $resolved_label -- the
// gap this comment used to describe as accepted is closed, not just documented.
function keySuffix(key) {
  if (!key) return '';
  // Model names can contain ':' (e.g. "qwen3.8:27b-q4_K_M") -- not filesystem-hostile on
  // Linux, but sanitized anyway so this never depends on a particular OS's path rules.
  return `.${String(key).replace(/[^A-Za-z0-9._-]+/g, '_')}`;
}

function lockFileName(key) {
  return `.pipeline-single-flight${keySuffix(key)}.lock`;
}

function lockFilePath(instancesDir, key) {
  return path.join(instancesDir, lockFileName(key));
}

// FIFO ticket queue directory (per model key, same suffix scheme as the lockfile).
function queueDirPath(instancesDir, key) {
  return path.join(instancesDir, `.pipeline-single-flight-queue${keySuffix(key)}`);
}

// 2026-08-24 (Grimmethy: "move the user interaction to the highest priority possible...
// if a user interaction shows up, it's next") -- same .discuss-waiting/ advisory
// directory agent-manager-common.sh's acquire_single_flight_lock() now checks (Python's
// single_flight_lock.py drops one marker file per waiting Discuss session in here),
// mirrored here so worker-1/worker-reasoning's OWN acquire (via draftTask()'s
// withLockFn, this module, not the bash function) also backs off instead of immediately
// reclaiming the lock for its next queued task the moment the current holder releases.
// Can't interrupt an in-flight call (and shouldn't -- that would waste real work), so
// this only affects who wins the race to acquire NEXT. A directory of per-waiter files,
// not one shared flag, so a second concurrent Discuss session's priority isn't silently
// cleared the instant the first one gets the lock.
const DISCUSS_PRIORITY_DIR_NAME = '.discuss-waiting';
const DISCUSS_PRIORITY_MAX_WAIT_MS = 8000;

function priorityWaitDirPath(instancesDir) {
  return path.join(instancesDir, DISCUSS_PRIORITY_DIR_NAME);
}

function someoneIsWaiting(instancesDir) {
  try {
    return fs.readdirSync(priorityWaitDirPath(instancesDir)).length > 0;
  } catch (e) {
    if (e.code === 'ENOENT') {
      return false; // directory doesn't exist yet -- nobody has ever waited.
    }
    console.error(`[single-flight-lock] someoneIsWaiting: unexpected error reading ${priorityWaitDirPath(instancesDir)}: code=${e.code} message=${e.message}`);
    throw e;
  }
}

// 2026-08-27, root-caused live: this used to be a genuinely unbounded blocking acquire
// ("no timeout, no -n, matching the bash version exactly" -- see agent-manager-common.sh's
// own matching fix for the full incident). Confirmed via `lslocks`: the kernel reported
// this exact lockfile held by a PID that no longer existed -- some prior holder died in a
// way that left the flock() unreleased, and every real caller across all three daemons
// (bash and Node both, since they share one kernel-level lock -- see this file's own
// header) sat blocked for 20+ minutes with zero recovery path, stalling the whole
// pipeline. `-w` bounds the wait via the same underlying `flock` binary the bash version
// uses (interoperable timeout semantics, not just interoperable locking); a real, non-
// stale holder finishing well within this window is unaffected. On timeout, `flock`
// exits non-zero, execFileSync throws, and this function's own catch below closes the fd
// and rethrows -- callers already going through withLock() get that exception the normal
// way, which is exactly what its "MUST be released -- an uncaught throw ... would leak
// the fd" header already documents as the expected failure contract.
// Read at call time, not module-load time (a test needs to override it per-call via
// process.env without needing a fresh require of this module each time).
function lockTimeoutSecs() {
  return Number(process.env.SINGLE_FLIGHT_LOCK_TIMEOUT_SECS) || 600;
}

// ---- per-process reentrancy ----------------------------------------------------------
// A lock this process already holds for a given key is re-granted immediately, without
// touching flock/the queue, and released only when the OUTERMOST holder releases (see
// header bug (1)). Keyed by the resolved lock FILE PATH so two different models are
// tracked independently. This is per-process state only: a different process asking for
// the same key still goes through the real ticket queue + flock, so cross-process mutual
// exclusion is unchanged.
const held = new Map(); // lockFilePath -> { fd, depth }

function reentrantToken(lockPath) {
  return { __sflReentrant: true, lockPath };
}

// Sub-second sync sleep (this whole module blocks the calling process by design).
function sleepMs(ms) {
  try {
    execFileSync('sleep', [(ms / 1000).toFixed(3)]);
  } catch (e) {
    // `sleep` missing / interrupted -- fall back to a busy wait so the loop still paces.
    const until = Date.now() + ms;
    while (Date.now() < until) { /* spin */ }
  }
}

function pidAlive(pid) {
  if (!Number.isInteger(pid) || pid <= 0) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (e) {
    return e.code === 'EPERM'; // exists but owned by someone else -- treat as alive
  }
}

// Remove tickets whose owning pid is gone, or that are older than the stale window (a
// leftover from a hard-killed process that never cleaned up). Best-effort: ENOENT races
// with another sweeper are ignored.
function sweepStaleTickets(qdir, staleMs, now) {
  let names;
  try {
    names = fs.readdirSync(qdir);
  } catch (e) {
    return;
  }
  for (const name of names) {
    const m = /^\d{20}\.(\d+)\./.exec(name);
    if (!m) continue; // not one of ours
    const full = path.join(qdir, name);
    let stale = !pidAlive(Number(m[1]));
    if (!stale) {
      try {
        stale = now - fs.statSync(full).mtimeMs > staleMs;
      } catch (e) {
        stale = true; // vanished under us -- nothing to keep
      }
    }
    if (stale) {
      try { fs.unlinkSync(full); } catch (e) { /* already gone */ }
    }
  }
}

function makeTimeoutError(key, timeoutSecs, detail) {
  // execFileSync's own error message is just "Command failed: flock -w 600 3" -- it
  // doesn't say WHY, and critically doesn't contain the word "timed out" that
  // local-worker.sh's/review-runner.sh's shared INFRA_FAILURE_PATTERN matches on to
  // decide "requeue and retry" vs "permanently block as a real content failure." A lock
  // timeout must read as infra-shaped, not as this task's own fault.
  return new Error(`single-flight lock acquisition timed out after ${timeoutSecs}s waiting for lock '${key || '(default)'}': ${detail}`);
}

const TICKET_POLL_MS = 500; // how often a not-yet-at-head waiter re-checks the queue (matches the 1s-ish granularity the discuss backoff already uses)
const FLOCK_ATTEMPT_SECS = 30; // per-attempt cap while at the head of the queue; loop until the overall deadline

// Blocking, exclusive acquire, bounded by SINGLE_FLIGHT_LOCK_TIMEOUT_SECS. Returns a
// handle to pass to release(): a real fd (number) for the outermost holder, or an opaque
// reentrant token if this process already holds the same key. Callers must not inspect
// the handle -- just pass it back to release().
function acquire(instancesDir, key) {
  const lp = lockFilePath(instancesDir, key);

  // Reentrant fast path -- this process already holds it (see header bug (1)).
  const existing = held.get(lp);
  if (existing) {
    existing.depth += 1;
    return reentrantToken(lp);
  }

  const timeoutSecs = lockTimeoutSecs();
  const overallDeadline = Date.now() + timeoutSecs * 1000;

  // Discuss priority: a user-interactive waiter (Python drops a marker in
  // .discuss-waiting/) gets a short head start to enter the queue before this lane does.
  const discussDeadline = Date.now() + DISCUSS_PRIORITY_MAX_WAIT_MS;
  while (someoneIsWaiting(instancesDir) && Date.now() < discussDeadline) {
    sleepMs(1000);
  }

  // Enqueue a FIFO ticket. Name = <20-digit ms><.pid><.rand> so a plain lexical sort is
  // chronological (all callers read the same wall clock -- localhost only).
  const qdir = queueDirPath(instancesDir, key);
  fs.mkdirSync(qdir, { recursive: true });
  const ticketName = `${String(Date.now()).padStart(20, '0')}.${process.pid}.${crypto.randomBytes(4).toString('hex')}`;
  const ticketPath = path.join(qdir, ticketName);
  fs.writeFileSync(ticketPath, '');

  const staleMs = timeoutSecs * 1000 * 2;
  try {
    for (;;) {
      if (Date.now() >= overallDeadline) {
        throw makeTimeoutError(key, timeoutSecs, 'never reached the head of the single-flight FIFO queue');
      }
      sweepStaleTickets(qdir, staleMs, Date.now());

      let headName;
      try {
        headName = fs.readdirSync(qdir).filter((n) => /^\d{20}\./.test(n)).sort()[0];
      } catch (e) {
        headName = ticketName; // dir vanished -- treat ourselves as first
      }
      if (headName && headName !== ticketName) {
        sleepMs(TICKET_POLL_MS);
        continue;
      }

      // We're at the head -- attempt the real flock. A non-participating bash/Python
      // caller (or a just-swept race) may still hold it; retry until the overall deadline.
      const remainingSecs = Math.max(1, Math.ceil((overallDeadline - Date.now()) / 1000));
      const attemptSecs = Math.min(FLOCK_ATTEMPT_SECS, remainingSecs);
      const fd = fs.openSync(lp, 'w');
      try {
        execFileSync('flock', ['-w', String(attemptSecs), String(LOCK_CHILD_FD)], { stdio: ['ignore', 'ignore', 'ignore', fd] });
      } catch (err) {
        fs.closeSync(fd);
        if (Date.now() >= overallDeadline) {
          throw makeTimeoutError(key, timeoutSecs, `held by another lane for the full wait (${err.message})`);
        }
        continue; // holder outlasted this attempt window -- re-check the queue and retry
      }
      held.set(lp, { fd, depth: 1 });
      return fd;
    }
  } finally {
    try { fs.unlinkSync(ticketPath); } catch (e) { /* already gone */ }
  }
}

// Releases a lock acquired above. Safe to call more than once, with null/undefined, or
// with an already-closed fd (best-effort, matching release_single_flight_lock()'s own
// `|| true`). For a reentrant token this just decrements the depth; the underlying fd is
// closed only when the outermost holder releases (depth hits 0).
function release(handle) {
  if (handle == null) return;

  if (typeof handle === 'object' && handle.__sflReentrant) {
    const e = held.get(handle.lockPath);
    if (e && e.depth > 0) e.depth -= 1;
    return;
  }

  // A raw fd (number) -- the outermost holder. Decrement its depth; close for real at 0.
  for (const [lp, e] of held) {
    if (e.fd === handle) {
      e.depth -= 1;
      if (e.depth <= 0) {
        try { fs.closeSync(e.fd); } catch (err) { /* already closed */ }
        held.delete(lp);
      }
      return;
    }
  }
  // Not tracked (already released, or an fd from a direct acquire predating this module's
  // reentrancy tracking) -- best-effort close, same as the old behavior.
  try { fs.closeSync(handle); } catch (err) { /* already closed */ }
}

// Preferred entry point: acquire, run fn, always release -- even if fn throws. Awaits fn
// (sync or async) and re-throws whatever it throws, after releasing.
async function withLock(instancesDir, fn, key) {
  const handle = acquire(instancesDir, key);
  try {
    return await fn();
  } finally {
    release(handle);
  }
}

// 2026-08-24 (Grimmethy: "add the lock work" for a live Discuss/worker-1 contention bug)
// -- python/dashboard/single_flight_lock.py is the dashboard-side twin of this file,
// letting Discuss's LOCAL provider join the exact same GPU/model-slot mutex worker-1 and
// reviewer already use. Deliberately did NOT add an equivalent lock for Claude calls
// (worker-reasoning / Discuss's Claude provider): unlike Ollama's single resident model,
// the Claude Code subscription has no real single-execution-slot constraint -- separate
// `claude -p` calls already run genuinely concurrently (the same way two independent
// Claude Code sessions do), bounded only by Anthropic's own account-level rate limits, not
// a local hardware bottleneck. Forcing them through a mutex here would have been a real
// regression, not a fix: a multi-minute adhoc draft (real Bash/Edit/Write investigation)
// would block a live chat, or vice versa, for a resource conflict that doesn't actually
// exist.
module.exports = { acquire, release, withLock, lockFilePath, queueDirPath };

File: python/dashboard/single_flight_lock.py
"""Python-side counterpart to src/single-flight-lock.js -- same real flock(2) mutex,
fully interoperable with the Node and bash versions (all three open the SAME lockfile
path; flock(2) locks are owned by the open file description, not the process, so any
combination of Python/Node/bash holding or waiting on it works correctly together).

Built 2026-08-24 after a live incident: clicking Discuss (local/Ornith provider) hit a
raw 500 because worker-1 was mid-draft on the same Ollama model at that exact moment --
Discuss had zero coordination with the worker lanes' own GPU-serialization discipline
(the .pipeline-single-flight.lock every Ollama-calling lane already uses, see
agent-manager-common.sh and single-flight-lock.js's own headers for why that lock
exists). The uncaught-timeout half of that incident was fixed separately in
discuss_sessions.py/app.py; this closes the actual resource-contention gap that caused
the timeout in the first place -- Discuss now waits its turn instead of racing a worker
lane for the one real GPU/model slot.

Only ONE lockfile, deliberately -- an earlier version of this also added a separate
.claude-single-flight.lock for the Claude Code subscription (worker-reasoning's adhoc
drafts / Discuss's Claude provider), on the theory that it needed the same coordination.
Reconsidered before shipping: unlike Ollama's single resident model, the Claude
subscription has no real single-execution-slot constraint -- separate `claude -p` calls
already run genuinely concurrently (same as two independent Claude Code sessions),
bounded only by Anthropic's own account-level rate limits, not a local hardware
bottleneck. Locking Claude calls against each other would have been a real regression,
not a fix: a multi-minute adhoc draft (real Bash/Edit/Write investigation) would block a
live chat, or vice versa, for a resource conflict that doesn't actually exist.

Priority for the human (2026-08-24, Grimmethy: "move the user interaction to the highest
priority possible... no matter what other tasks may be in queue, if a user interaction
shows up, it's next"): fairness alone still wasn't enough -- caught live, a Discuss
session lost race after race against worker-1 continuously reclaiming the lock for its
next queued task, and sat for minutes before ultimately failing. A held flock genuinely
can't be interrupted mid-call (and shouldn't be -- killing a half-finished generation
wastes real work), so this can't preempt whichever call is CURRENTLY running; what it CAN
do is stop a worker/reviewer lane from immediately racing to reclaim the lock for its
NEXT task the instant the current holder releases. held() below drops a per-waiter marker
file into instances/.discuss-waiting/ before blocking on flock, and removes it the moment
it actually acquires -- agent-manager-common.sh's acquire_single_flight_lock() and
single-flight-lock.js's acquire() both check whether that directory is non-empty and back
off a few seconds before attempting their own acquire, real headroom for Discuss's
already-parked blocking flock() call to win the wakeup race. A DIRECTORY of per-waiter
files, not one shared flag, so a second concurrent Discuss session doesn't have its
priority silently cleared the instant the first one gets the lock.

Split into acquire()/release() (2026-08-24, Chat panel's "fully reserve the reasoning
model" toggle) plus held(), a thin contextmanager wrapper around them -- a reservation has
to span multiple separate HTTP requests (toggle on, several chat messages, toggle off),
which a single `with` block can't do. held() stays the right shape for every OTHER caller
(Discuss, and Chat's own per-message calls when NOT reserved), where the lock's whole
life fits inside one function call.
"""
import fcntl
import os
import time
import uuid
from pathlib import Path
from contextlib import contextmanager

LOCK_NAME = ".pipeline-single-flight.lock"
PRIORITY_WAIT_DIR_NAME = ".discuss-waiting"


def _timeout_secs() -> float:
    """Bounded-wait parity with single-flight-lock.js / agent-manager-common.sh
    (2026-08-31). Both of those bound acquire() with `flock -w $SINGLE_FLIGHT_LOCK_TIMEOUT_SECS`
    (default 600) -- this twin was still an *unbounded* fcntl.flock(LOCK_EX), so a stuck
    holder (a dead PID that left the flock unreleased -- the exact 2026-08-27 incident
    the other two twins were bounded for) would hang a dashboard request forever. Same
    env var, same default."""
    try:
        return float(os.environ.get("SINGLE_FLIGHT_LOCK_TIMEOUT_SECS") or 600)
    except (TypeError, ValueError):
        return 600.0


def _lock_name(key: str | None) -> str:
    """Per-model locking (2026-08-25 -- see src/single-flight-lock.js's matching header
    for the full incident and reasoning; this mirrors it exactly so a Python caller and a
    JS caller hitting the SAME model still share the SAME lockfile). `key` is normally
    ollama_client.MODEL (== LOCAL_MODEL) -- the model this call will actually hit. Omitted
    (None), this returns the original global LOCK_NAME unchanged, so any caller not
    passing a key keeps today's behavior exactly."""
    if not key:
        return LOCK_NAME
    import re
    safe_key = re.sub(r"[^A-Za-z0-9._-]+", "_", str(key))
    return f".pipeline-single-flight.{safe_key}.lock"


def acquire(instances_dir: Path, key: str | None = None):
    """Blocking, exclusive acquire, bounded by SINGLE_FLIGHT_LOCK_TIMEOUT_SECS (default
    600) -- parity with single-flight-lock.js / agent-manager-common.sh's `flock -w`.
    Drops a per-waiter priority marker (see this module's own header) for the duration of
    the wait, clearing it the instant the lock is actually held. Returns the open file
    object -- it MUST stay open for as long as the lock should be held (the flock is owned
    by the open file description, not the process) and MUST be passed to release() when
    done, or the lock leaks for the life of the process. Raises TimeoutError (message
    contains "timed out", matching the JS twin's INFRA_FAILURE_PATTERN-shaped string) if
    the wait deadline elapses."""
    instances_dir.mkdir(parents=True, exist_ok=True)
    wait_dir = instances_dir / PRIORITY_WAIT_DIR_NAME
    wait_dir.mkdir(exist_ok=True)
    marker = wait_dir / uuid.uuid4().hex
    marker.touch()
    try:
        lock_path = instances_dir / _lock_name(key)
        fh = open(lock_path, "w")
        timeout = _timeout_secs()
        deadline = time.monotonic() + timeout
        while True:
            try:
                fcntl.flock(fh, fcntl.LOCK_EX | fcntl.LOCK_NB)
                return fh
            except OSError:
                if time.monotonic() >= deadline:
                    fh.close()
                    raise TimeoutError(
                        f"single-flight lock acquisition timed out after {timeout:.0f}s "
                        f"waiting for lock '{key or '(default)'}'"
                    )
                time.sleep(0.25)
    finally:
        # Acquired (or the open/flock call itself raised) -- either way no longer
        # "waiting". Not everyone's marker: another concurrent waiter may still be queued
        # behind us and its own priority claim must stay intact.
        marker.unlink(missing_ok=True)


def release(fh):
    """Releases a lock acquired above. Safe to call with None (best-effort, matching
    single-flight-lock.js's own release() semantics)."""
    if fh is None:
        return
    try:
        fcntl.flock(fh, fcntl.LOCK_UN)
    finally:
        fh.close()


@contextmanager
def held(instances_dir: Path, key: str | None = None):
    """Preferred entry point when the lock's whole life fits inside one function call --
    callers should wrap just the real model call, not any surrounding prompt-building/
    grep work, same discipline local-draft.js's own withLock() usage already follows.
    Pass `key` (normally ollama_client.MODEL) so this shares a lockfile with every other
    caller -- JS or Python -- hitting the SAME model; omitted, this locks against the
    original global lockfile."""
    fh = acquire(instances_dir, key)
    try:
        yield
    finally:
        release(fh)
